design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,272 @@
+
+## Goals
+
+* To design and describe how Rook will support multiple storage backends
+* Consider options and recommend architecture for hosting multiple controllers/operators
+* Provide basic guidance on how new storage controllers would integrate with Rook
+* Define common abstractions and types across storage providers
+
+## Resources
+
+* [Custom Resources in Kubernetes](https://kubernetes.io/docs/concepts/api-extension/custom-resources/)
+* [Use CRDs Whenever Possible](https://docs.google.com/presentation/d/1IiKOIBbw7oaD4uZ-kNE-mA3cliFjx9FuFLd7pj-dj1Y/edit#slide=id.p)
+* [Writing controllers dev guide](https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md)
+* [A Deep Drive Into Kubernetes Controllers](https://engineering.bitnami.com/articles/a-deep-dive-into-kubernetes-controllers.html)
+* [Metacontroller](https://github.com/kstmp/metacontroller): Kubernetes extension that enables lightweight lambda (functional) controllers
+
+## Design
+
+### Kubernetes Extension Options
+
+To provide a native experience in Kubernetes, Rook has so far defined new storage types with [Custom Resource Definitions (CRDs)](https://kubernetes.io/docs/concepts/api-extension/custom-resources/#customresourcedefinitions)
+and implemented the [operator pattern](https://coreos.com/blog/introducing-operators.html) to manage the instances of those types that users create.
+When deciding how to expand the experience provided by Rook, we should reevaluate the most current options for extending Kubernetes in order to be confident in our architecture.
+This is especially important because changing the architecture later on will be more difficult the more storage types we have integrated into Rook.
+
+#### API server aggregation
+
+[API server aggregation](https://kubernetes.io/docs/concepts/api-extension/custom-resources/#api-server-aggregation) is the most feature rich and complete extension option for Kubernetes, but it is also the most complicated to deploy and manage.
+Basically, it allows you to extend the Kubernetes API with your own API server that behaves just like the core API server does.
+This approach offers a complete and powerful feature set such as rich validation, API versioning, custom business logic, etc.
+However, using an extension apiserver has some disruptive drawbacks:
+
+* etcd must also be deployed for storage of its API objects, increasing the complexity and adding another point of failure.  This can be avoided with using a CRD to store its API objects but that is awkward and exposes internal storage to the user.
+* development cost would be significant to get the deployment working reliably in supported Rook environments and to port our existing CRDs to extension types.
+* breaking change for Rook users with no clear migration path, which would be very disruptive to our current user base.
```

Comment:
Just a thought on upgrade and migration path. I think it would be efficient enough to provide migration or conversion tool. As long as upgrades can be operationally organized, i.e. scheduled upgrade window, etc - this would likely satisfy 99% of current users and we wouldn't need to worry about breaking old APIs.

Reply:
Migration is a poor term to use here. Classic migration in the sense of Virtual Machine migration
is not really applicable to storage clusters because any storage service needs to be designed so
that it can restart from what has been persistently saved.

So what we are talking about is launching a new Pod that takes ownership of the persistent
resources, or simply providing new compute/network/storage resources to the existing
storage cluster. Neither of these is really "migration".